src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+            head = tail = new Node(null, value, null);
+        } else {
+            Node notTheTailAnyMore = tail;
+            tail = new Node(notTheTailAnyMore, value, null);
+            notTheTailAnyMore.next = tail;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            if (index == size) {
+                add(value);
+            } else if (index == 0) {
+                Node notTheHeadAnyMore = head;
+                head = new Node(null, value, notTheHeadAnyMore);
+                notTheHeadAnyMore.prev = head;
+                size++;
+            } else {
+                int currentIndex = 0;
+                if (size > 1 && index > (int) (size / 2)) {
+                    currentIndex = (int) (size / 2) - 1;
+                }
+                Node currentNode = head;
+                while (currentIndex != index) {
+                    currentNode = currentNode.next;
+                    currentIndex++;
+                }
+                Node newNode = new Node(currentNode.prev, value, currentNode);
+                currentNode.prev.next = newNode;
+                currentNode.prev = newNode;
+
+                size++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
+    }
+
+    public Node getNode(int index) {
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            if (index < (int) (size / 2)) {
+                int currentIndex = 0;
+                Node currentNode = head;
+                while (currentIndex != index) {
+                    currentNode = currentNode.next;
+                    currentIndex++;
+                }
+                return (Node<T>) currentNode;
+            } else  {
+                int currentIndex = size - 1;
+                Node currentNode = tail;
+                while (currentIndex != index) {
+                    currentNode = currentNode.prev;
+                    currentIndex--;
+                }
+                return (Node<T>) currentNode;
+            }
+
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return (T) getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            T previousNode = (T) getNode(index).value;
+            getNode(index).value = value;
+            return previousNode;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
```

Comment:
Common mistake: Don't complicate if-else construction. In this case, you should remove the `else` clause because it's redundant. If the condition in the `if` clause is true, the method will exit due to the `throw` statement.